code/controllers/subsystem/xenoarchaeology.dm

```diff
@@ -0,0 +1,241 @@
+	var/list/xenoartifact_trait_priorities = list(TRAIT_PRIORITY_ACTIVATOR, TRAIT_PRIORITY_MINOR, TRAIT_PRIORITY_MALFUNCTION, TRAIT_PRIORITY_MAJOR)
+
+	///List of 'discovered' traits
+	var/list/discovered_traits = list()
+
+/datum/controller/subsystem/xenoarchaeology/Initialize(timeofday)
+	. = ..()
+	//Poplate seller & artifact personalities
+	xenoa_seller_names = world.file2list("strings/names/science_seller.txt")
+	xenoa_seller_dialogue = world.file2list("strings/science_dialogue.txt")
+	xenoa_artifact_names = world.file2list("strings/names/artifact_sentience.txt")
+
+	//Dirty unwashed masses
+	xenoa_all_traits = compile_artifact_weights(/datum/xenoartifact_trait)
+	xenoa_all_traits_keyed = compile_artifact_weights(/datum/xenoartifact_trait, TRUE)
+
+	//Compatabilities
+	xenoa_item_incompatible = compile_artifact_compatibilties(TRAIT_INCOMPATIBLE_ITEM)
+	xenoa_mob_incompatible = compile_artifact_compatibilties(TRAIT_INCOMPATIBLE_MOB)
+	xenoa_structure_incompatible = compile_artifact_compatibilties(TRAIT_INCOMPATIBLE_STRUCTURE)
+
+	//Labeler
+	labeler_traits = new()
+	labeler_traits.compile_artifact_whitelist(/datum/xenoartifact_material)
+	labeler_traits_filter = list()
+	for(var/datum/xenoartifact_trait/T as() in xenoa_all_traits)
+		T = new T() //Instantiate so we can access a PROC
+		var/list/hints = T.get_dictionary_hint()
+		for(var/i in hints)
+			if(!labeler_traits_filter[i["icon"]])
+				labeler_traits_filter[i["icon"]] = list()
+			labeler_traits_filter[i["icon"]] += list("[initial(T.label_name)]")
+		QDEL_NULL(T)
+
+	//Populate traits by material
+	material_traits = list()
+	for(var/datum/xenoartifact_material/material_index as() in typesof(/datum/xenoartifact_material))
+		if(SSxenoarchaeology.material_traits[initial(material_index.material_parent)])
+			continue
+		var/datum/xenoa_material_traits/material = new()
+		material_traits[material_index] = material
+		//Populate datum fields
+		material.compile_artifact_whitelist(material_index)
+
+/datum/controller/subsystem/xenoarchaeology/Shutdown()
+	. = ..()
+
+/datum/controller/subsystem/xenoarchaeology/Recover()
+	. = ..()
+	xenoa_seller_names = SSxenoarchaeology.xenoa_seller_names
+	xenoa_seller_dialogue = SSxenoarchaeology.xenoa_seller_dialogue
+	xenoa_artifact_names = SSxenoarchaeology.xenoa_artifact_names
+
+	xenoa_all_traits = SSxenoarchaeology.xenoa_all_traits
+	xenoa_all_traits_keyed = SSxenoarchaeology.xenoa_all_traits_keyed
+
+	material_traits = SSxenoarchaeology.material_traits
+
+	xenoa_item_incompatible = SSxenoarchaeology.xenoa_item_incompatible
+	xenoa_mob_incompatible = SSxenoarchaeology.xenoa_mob_incompatible
+	xenoa_structure_incompatible = SSxenoarchaeology.xenoa_structure_incompatible
+
+	labeler_tooltip_stats = SSxenoarchaeology.labeler_tooltip_stats
+	labeler_traits_filter = SSxenoarchaeology.labeler_traits_filter
+
+/datum/controller/subsystem/xenoarchaeology/proc/register_console(var/obj/machinery/computer/xenoarchaeology_console/new_console)
```

Comment:
remove `var` from the parameter. `obj/...` is enough here.

Reply:
True